department-of-veterans-affairs/caseflow
# Stuck Appeals Job now finds appeals where on_hold tasks do not have an active descendant

Resolves #15267
 
### Description
Extends the Stuck Appeals Job, All On Hold query to check for appeals with a fully on hold subtree.

Extended the existing query in order to avoid duplication of found appeals.

### Acceptance Criteria
- [ ] regularly running job that reports on on_hold tasks that do not have an active descendant task

### Testing Plan
Console based!

 1. Confirm only six database calls are made when invoking `AppealsWithNoTasksOrAllTasksOnHoldQuery` (five existing, one new)
 1. Check the right appeals are found, the wrong aren't:

#### Appeals that should be caught

Appeal has an on_hold task with no children:
```ruby
appeal = FactoryBot.create(:appeal, :with_post_intake_tasks)                                                                   
task = FactoryBot.create(:privacy_act_task, appeal: appeal, parent: appeal.root_task, assigned_by: User.first, assigned_to: User.last)                                          
task.descendants.each(&:on_hold!)                                                                                   
AppealsWithNoTasksOrAllTasksOnHoldQuery.new.ama_appeal_stuck?(appeal)
# true
```
```
                                         ┌────────────────────────────────────────────────────────────────────────┐
Appeal 1065 (evidence_submission) ────── │ ID   │ STATUS    │ ASGN_BY     │ ASGN_TO     │ UPDATED_AT              │
└── RootTask                             │ 3488 │ on_hold   │             │ Bva         │ 2020-11-10 20:42:59 UTC │
    ├── DistributionTask                 │ 3489 │ on_hold   │             │ Bva         │ 2020-11-10 20:42:59 UTC │
    │   └── EvidenceSubmissionWindowTask │ 3490 │ assigned  │             │ MailTeam    │ 2020-11-10 20:42:59 UTC │
    └── PrivacyActTask                   │ 3491 │ on_hold   │ BVASCASPER1 │ BVADPACOCHA │ 2020-11-10 20:43:34 UTC │
                                         └────────────────────────────────────────────────────────────────────────┘
```

Appeal has an on hold task with only closed children:
```ruby
appeal = FactoryBot.create(:appeal, :with_post_intake_tasks)                                                                   
task = FactoryBot.create(:privacy_act_task, appeal: appeal, parent: appeal.root_task, assigned_by: User.first, assigned_to: User.last)                                          
task.descendants.each(&:on_hold!)  
task2 = FactoryBot.create(:privacy_act_task, appeal: appeal, parent: task, assigned_by: User.first, assigned_to: User.last)                                          
task2.descendants.each(&:completed!)  
AppealsWithNoTasksOrAllTasksOnHoldQuery.new.ama_appeal_stuck?(appeal)
# true
```
```
                                         ┌────────────────────────────────────────────────────────────────────────┐
Appeal 1065 (evidence_submission) ────── │ ID   │ STATUS    │ ASGN_BY     │ ASGN_TO     │ UPDATED_AT              │
└── RootTask                             │ 3488 │ on_hold   │             │ Bva         │ 2020-11-10 20:42:59 UTC │
    ├── DistributionTask                 │ 3489 │ on_hold   │             │ Bva         │ 2020-11-10 20:42:59 UTC │
    │   └── EvidenceSubmissionWindowTask │ 3490 │ assigned  │             │ MailTeam    │ 2020-11-10 20:42:59 UTC │
    └── PrivacyActTask                   │ 3491 │ on_hold   │ BVASCASPER1 │ BVADPACOCHA │ 2020-11-10 20:43:34 UTC │
        └── PrivacyActTask               │ 3492 │ completed │ BVASCASPER1 │ BVADPACOCHA │ 2020-11-10 20:43:54 UTC │
                                         └────────────────────────────────────────────────────────────────────────┘
```

#### Appeals that should _not_ be caught:

Appeal has an on hold task with open children:
```ruby
appeal = FactoryBot.create(:appeal, :with_post_intake_tasks)
AppealsWithNoTasksOrAllTasksOnHoldQuery.new.ama_appeal_stuck?(appeal)
# false
```
```
                                         ┌────────────────────────────────────────────────────────────────────────┐
Appeal 1065 (evidence_submission) ────── │ ID   │ STATUS    │ ASGN_BY     │ ASGN_TO     │ UPDATED_AT              │
└── RootTask                             │ 3488 │ on_hold   │             │ Bva         │ 2020-11-10 20:42:59 UTC │
    ├── DistributionTask                 │ 3489 │ on_hold   │             │ Bva         │ 2020-11-10 20:42:59 UTC │
        └── EvidenceSubmissionWindowTask │ 3490 │ assigned  │             │ MailTeam    │ 2020-11-10 20:42:59 UTC │
                                         └────────────────────────────────────────────────────────────────────────┘
```

Appeal has an on hold task with closed and open children:
```ruby
appeal = FactoryBot.create(:appeal, :with_post_intake_tasks)                                                                   
task = FactoryBot.create(:privacy_act_task, appeal: appeal, parent: DistributionTask.find_by(appeal: appeal), assigned_by: User.first, assigned_to: User.last)                                          
task.descendants.each(&:completed!)  
AppealsWithNoTasksOrAllTasksOnHoldQuery.new.ama_appeal_stuck?(appeal)
# false
```
```
                                         ┌────────────────────────────────────────────────────────────────────────┐
Appeal 1065 (evidence_submission) ────── │ ID   │ STATUS    │ ASGN_BY     │ ASGN_TO     │ UPDATED_AT              │
└── RootTask                             │ 3488 │ on_hold   │             │ Bva         │ 2020-11-10 20:42:59 UTC │
    ├── DistributionTask                 │ 3489 │ on_hold   │             │ Bva         │ 2020-11-10 20:42:59 UTC │
        └── EvidenceSubmissionWindowTask │ 3490 │ assigned  │             │ MailTeam    │ 2020-11-10 20:42:59 UTC │
        └── PrivacyActTask               │ 3492 │ completed  │ BVASCASPER1 │ BVADPACOCHA │ 2020-11-10 20:43:54 UTC │

                                         └────────────────────────────────────────────────────────────────────────┘
```
